compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/CallGenerator.kt

```diff
@@ -128,8 +130,15 @@ class CallGenerator(statementGenerator: StatementGenerator) : StatementGenerator
                 context.callToSubstitutedDescriptorMap[this] = getterDescriptor
                 putTypeArguments(typeArguments) { it.toIrType() }
             }
-        } else
-            IrGetValueImpl(startOffset, endOffset, descriptor.type.toIrType(), context.symbolTable.referenceValue(descriptor), origin)
+        } else {
+            val getValue =
```

Comment:
Please, add irText test case(s) for issues with non-optimized `when` you had in mind.

Reply:
Added as `whenSmartCastToEnum`. The change is that the temporary variable now has type `En` and not `Any` or `Any?`. Then the enum optimizations will take it from there to turn it into a switch.